stat.c

```diff
@@ -3087,17 +3091,29 @@ int calc_log_samples(void)
 		if (!td->bw_log ||
 			(td->bw_log && !per_unit_log(td->bw_log))) {
 			tmp = add_bw_samples(td, &now);
-			if (tmp < next)
-				next = tmp;
+
+			if (td->bw_log)
+				log_avg_msec = td->bw_log->avg_msec;
 		}
 		if (!td->iops_log ||
 			(td->iops_log && !per_unit_log(td->iops_log))) {
 			tmp = add_iops_samples(td, &now);
-			if (tmp < next)
-				next = tmp;
+
+			if (td->iops_log)
+				log_avg_msec = td->iops_log->avg_msec;
 		}
+
+		if (tmp < next)
+			next = tmp;
 	}
 
+	if (log_avg_msec == 0)
+		next_mod = elapsed_time;
+	else
+		next_mod = elapsed_time % log_avg_msec;
+
+	next = min(next, (log_avg_msec - next_mod));  /* correction to keep the time on the log avg msec boundary */
```

Comment:
So `log_avg_msec - next_mod` is depending on wraparound to make it large in the `log_avg_msec` case? This feels a bit strange...

Reply:
next_mod is the actual time expired in the interval and (log_avg_msec - next_mod)  is the actual time left in interval.  So when next_mod is 0 (elapsed_time is an even multiple of log_avg_msec), it indicates it's time to log the IOPs and BW and start a new interval.  Since it's based on elapsed time now, it will generate more accurate next value.  Make sense?

Reply:
Yup!